File: 剑指offer/32 - III. 从上到下打印二叉树 III.py
#######方法一最优
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root: return []
        res, queue = [], collections.deque()
        queue.append(root)
        while queue:
            tmp = []
            for _ in range(len(queue)):
                node = queue.popleft()
                tmp.append(node.val)
                if node.left: queue.append(node.left)
                if node.right: queue.append(node.right)
            res.append(tmp[::-1] if len(res) % 2 else tmp)
        return res
#时间复杂度：O(N)
#空间复杂度：O(N)
    

##########方法2 
class Solution:
    from collections import deque
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root: return []
        deque, res = collections.deque([root]), []
        while deque:
            tmp = collections.deque()
            for _ in range(len(deque)):
                node = deque.popleft()
                if len(res)%2: tmp.appendleft(node.val)
                else: tmp.append(node.val)
                if node.left: deque.append(node.left)
                if node.right: deque.append(node.right)
            res.append(list(tmp))
        return res
    


###至于倒数第二行为何要加list(tmp)，看以下代码即懂
from collections import deque
tmp = deque()
tmp.appendleft([1, 2, 3])
tmp.append([1, 2, 3])
print(tmp)         #返回deque([[1, 2, 3]])
print(list(tmp))   #返回[[1, 2, 3]]

#############方法3 拆开分别考虑
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root: return []
        res, deque = [], collections.deque()
        deque.append(root)
        while deque:
            tmp = []
            # 打印奇数层
            for _ in range(len(deque)):
                # 从左向右打印
                node = deque.popleft()
                tmp.append(node.val)
                # 先左后右加入下层节点
                if node.left: deque.append(node.left)
                if node.right: deque.append(node.right)
            res.append(tmp)
            if not deque: break
            # 打印偶数层
            tmp = []
            for _ in range(len(deque)):
                # 从右向左打印
                node = deque.pop()
                tmp.append(node.val)
                # 先右后左加入下层节点
                if node.right: deque.appendleft(node.right)
                if node.left: deque.appendleft(node.left)
            res.append(tmp)
        return res

https://leetcode-cn.com/problems/cong-shang-dao-xia-da-yin-er-cha-shu-iii-lcof/solution/mian-shi-ti-32-iii-cong-shang-dao-xia-da-yin-er--3/

File: 剑指offer/C++_Version/32 - III. 从上到下打印二叉树 III.cpp
//方法1
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) 
    {
        vector<vector<int>> res;
        if(root==nullptr) return res;
        queue<TreeNode*> que;
        que.push(root);
        while(!que.empty())
        {
            int sz = que.size();
            vector<int> tmp;
            for(int i=0; i<sz; i++)
            {
                TreeNode *node = que.front();
                que.pop();
                tmp.push_back(node->val);
                if(node->left) que.push(node->left);
                if(node->right) que.push(node->right);
            }
            if(res.size()&1)
            {
                reverse(tmp.begin(), tmp.end());
                res.push_back(tmp);
            }
            else
                res.push_back(tmp);
        }
        return res;
    }
};
